java/src/main/java/actorworkflow/CustomerLoyaltyWorkflowImpl.java

```diff
@@ -0,0 +1,147 @@
+package actorworkflow;
+
+import io.temporal.activity.ActivityOptions;
+import io.temporal.api.common.v1.WorkflowExecution;
+import io.temporal.api.enums.v1.ParentClosePolicy;
+import io.temporal.client.WorkflowExecutionAlreadyStarted;
+import io.temporal.workflow.*;
+import org.slf4j.Logger;
+
+import java.time.Duration;
+import java.util.ArrayList;
+
+public class CustomerLoyaltyWorkflowImpl implements CustomerLoyaltyWorkflow {
+
+    private static final Logger logger = Workflow.getLogger(CustomerLoyaltyWorkflowImpl.class);
+
+    private final CustomerLoyaltyActivities activities =
+            Workflow.newActivityStub(
+                    CustomerLoyaltyActivities.class,
+                    ActivityOptions.newBuilder()
+                            .setStartToCloseTimeout(Duration.ofSeconds(2))
+                            .build());
+
+    boolean accountActive = true;
+
+    Customer customer;
+
+    @Override
+    public String customerLoyalty(Customer customer) {
+        System.out.println("Customer passed in is: " + customer);
+        this.customer = customer;
+        WorkflowInfo info = Workflow.getInfo();
+
+        logger.info("Started workflow. Customer: {}.", this.customer);
+
+        if (info.getContinuedExecutionRunId().isEmpty()) {
+            String tier = this.customer.status().name();
+            activities.sendEmail(EmailStrings.EMAIL_WELCOME.formatted(tier));
+        }
+
+        // block on everything
+        while (true) {
+            Workflow.await(() -> !accountActive || info.getHistoryLength() > Shared.HISTORY_THRESHOLD);
+            if (accountActive) {
+                logger.info("Account still active, history size crossed limit; continuing-as-new.");
+                Workflow.continueAsNew(this.customer);
+            } else {
+                logger.info("Account canceled. Closing workflow.");
+                return "Done";
+            }
+        }
+    }
+
+    @Override
+    public void addLoyaltyPoints(int pointsToAdd) {
+        customer = customer.withPoints(customer.loyaltyPoints() + pointsToAdd);
+        logger.info("Added {} points to customer. Total loyalty points now {}.",
+                pointsToAdd, customer.loyaltyPoints());
+
+        StatusTier tierToPromoteTo = StatusTier.getMaxTier(customer.loyaltyPoints());
```

Comment:
if i understand right these calculations are hard-coded but imho in most cases this would be a call to a service
that is responsible to calculate customer status / tier / etc?

Reply:
I could see it either way, but in my mind for this example, having the logic in the workflow makes sense because (a) it's deterministic (and fast); and (b) it's easier to read, follow, and understand (thus demonstrating the benefit of a workflow containing all of the logic for this use-case).

Reply:
i think there is benefits having this as independent service, for example you can update how you calculate loyalty points without having to change workflow code.
this code might be fast but note it will be executed each time worker needs to replay this execution.
in this case you are calculating this multiple times potentially. if you have to keep this inside workflow code i would at least change it to be executed as side effect so it would not be executed during replay